File: apple/VocabKit/Tests/VocabKitTests/SRSEngineTests.swift
import XCTest
@testable import VocabKit

final class SRSEngineTests: XCTestCase {
    let now = Date(timeIntervalSince1970: 1_700_000_000)

    private func daysBetween(_ state: SRSState, from: Date) -> Double {
        (state.dueDateValue.timeIntervalSince1970 - from.timeIntervalSince1970)
            / SRSEngine.secondsPerDay
    }

    func testFirstGoodReviewSchedulesOneDay() {
        let s = SRSEngine.schedule(.new(now: now), grade: .good, now: now)
        XCTAssertEqual(s.repetitions, 1)
        XCTAssertEqual(s.intervalDays, 1)
        XCTAssertEqual(s.easeFactor, 2.5, accuracy: 0.0001) // "good" leaves EF unchanged
        XCTAssertEqual(daysBetween(s, from: now), 1, accuracy: 0.01)
    }

    func testSecondGoodReviewSchedulesSixDays() {
        var s = SRSEngine.schedule(.new(now: now), grade: .good, now: now)
        s = SRSEngine.schedule(s, grade: .good, now: now)
        XCTAssertEqual(s.repetitions, 2)
        XCTAssertEqual(s.intervalDays, 6)
    }

    func testThirdGoodReviewUsesEaseFactor() {
        var s = SRSEngine.schedule(.new(now: now), grade: .good, now: now)
        s = SRSEngine.schedule(s, grade: .good, now: now)
        s = SRSEngine.schedule(s, grade: .good, now: now)
        XCTAssertEqual(s.repetitions, 3)
        XCTAssertEqual(s.intervalDays, 15) // round(6 * 2.5)
    }

    func testEasyRaisesEaseFactor() {
        let s = SRSEngine.schedule(.new(now: now), grade: .easy, now: now)
        XCTAssertEqual(s.easeFactor, 2.6, accuracy: 0.0001)
    }

    func testHardLowersEaseFactor() {
        let s = SRSEngine.schedule(.new(now: now), grade: .hard, now: now)
        XCTAssertEqual(s.easeFactor, 2.36, accuracy: 0.0001)
    }

    func testAgainResetsRepetitionsAndInterval() {
        var s = SRSEngine.schedule(.new(now: now), grade: .good, now: now)
        s = SRSEngine.schedule(s, grade: .good, now: now) // interval 6, reps 2
        s = SRSEngine.schedule(s, grade: .again, now: now)
        XCTAssertEqual(s.repetitions, 0)
        XCTAssertEqual(s.intervalDays, 1)
    }

    func testEaseFactorClampedAtMinimum() {
        var s = SRSState.new(now: now)
        for _ in 0..<10 {
            s = SRSEngine.schedule(s, grade: .again, now: now)
        }
        XCTAssertGreaterThanOrEqual(s.easeFactor, SRSEngine.minimumEase)
        XCTAssertEqual(s.easeFactor, SRSEngine.minimumEase, accuracy: 0.0001)
    }

    func testIsDue() {
        let past = SRSState(dueDate: SRSEngine.millis(now.addingTimeInterval(-3600)))
        let future = SRSState(dueDate: SRSEngine.millis(now.addingTimeInterval(3600)))
        XCTAssertTrue(SRSEngine.isDue(past, now: now))
        XCTAssertFalse(SRSEngine.isDue(future, now: now))
    }
}

File: windows/Vocb.Core.Tests/SrsEngineTests.cs
using Vocb.Core;
using Xunit;

namespace Vocb.Core.Tests;

public class SrsEngineTests
{
    // Fixed reference instant (matches the Swift tests' 1_700_000_000 seconds).
    private const long Now = 1_700_000_000_000L;

    private static double DaysUntilDue(SrsState s, long fromMs)
        => (s.DueDate - fromMs) / (double)SrsEngine.MillisPerDay;

    [Fact]
    public void FirstGoodReview_SchedulesOneDay()
    {
        var s = SrsEngine.Schedule(SrsState.New(Now), ReviewGrade.Good, Now);
        Assert.Equal(1, s.Repetitions);
        Assert.Equal(1, s.IntervalDays);
        Assert.Equal(2.5, s.EaseFactor, 4); // "good" leaves EF unchanged
        Assert.Equal(1, DaysUntilDue(s, Now), 3);
    }

    [Fact]
    public void SecondGoodReview_SchedulesSixDays()
    {
        var s = SrsEngine.Schedule(SrsState.New(Now), ReviewGrade.Good, Now);
        s = SrsEngine.Schedule(s, ReviewGrade.Good, Now);
        Assert.Equal(2, s.Repetitions);
        Assert.Equal(6, s.IntervalDays);
    }

    [Fact]
    public void ThirdGoodReview_UsesEaseFactor()
    {
        var s = SrsEngine.Schedule(SrsState.New(Now), ReviewGrade.Good, Now);
        s = SrsEngine.Schedule(s, ReviewGrade.Good, Now);
        s = SrsEngine.Schedule(s, ReviewGrade.Good, Now);
        Assert.Equal(3, s.Repetitions);
        Assert.Equal(15, s.IntervalDays); // round(6 * 2.5)
    }

    [Fact]
    public void Easy_RaisesEaseFactor()
    {
        var s = SrsEngine.Schedule(SrsState.New(Now), ReviewGrade.Easy, Now);
        Assert.Equal(2.6, s.EaseFactor, 4);
    }

    [Fact]
    public void Hard_LowersEaseFactor()
    {
        var s = SrsEngine.Schedule(SrsState.New(Now), ReviewGrade.Hard, Now);
        Assert.Equal(2.36, s.EaseFactor, 4);
    }

    [Fact]
    public void Again_ResetsRepetitionsAndInterval()
    {
        var s = SrsEngine.Schedule(SrsState.New(Now), ReviewGrade.Good, Now);
        s = SrsEngine.Schedule(s, ReviewGrade.Good, Now); // interval 6, reps 2
        s = SrsEngine.Schedule(s, ReviewGrade.Again, Now);
        Assert.Equal(0, s.Repetitions);
        Assert.Equal(1, s.IntervalDays);
    }

    [Fact]
    public void EaseFactor_ClampedAtMinimum()
    {
        var s = SrsState.New(Now);
        for (int i = 0; i < 10; i++)
            s = SrsEngine.Schedule(s, ReviewGrade.Again, Now);
        Assert.Equal(SrsEngine.MinimumEase, s.EaseFactor, 4);
    }

    [Fact]
    public void IsDue_ComparesDueDateToNow()
    {
        var past = new SrsState { DueDate = Now - 3_600_000 };
        var future = new SrsState { DueDate = Now + 3_600_000 };
        Assert.True(SrsEngine.IsDue(past, Now));
        Assert.False(SrsEngine.IsDue(future, Now));
    }
}
